File: src/merge-sort/merge-sort.js
#!/usr/bin/env node

function quickSort(arr) {
  if (arr.length <= 1) return arr;
  const pivot = arr[Math.floor(arr.length / 2)];
  const left = arr.filter(x => x < pivot);
  const mid = arr.filter(x => x === pivot);
  const right = arr.filter(x => x > pivot);
  return [...quickSort(left), ...mid, ...quickSort(right)];
}

function mergeSort(arr) {
  if (arr.length <= 1) return arr;
  const mid = Math.floor(arr.length / 2);
  const left = mergeSort(arr.slice(0, mid));
  const right = mergeSort(arr.slice(mid));
  return merge(left, right);
}

function merge(left, right) {
  const result = [];
  while (left.length && right.length)
    result.push(left[0] <= right[0] ? left.shift() : right.shift());
  return result.concat(left, right);
}

const nums = process.argv.slice(2).map(Number);
if (nums.length === 0) {
  console.error("Usage: node sort.js 5 2 9 1");
  process.exit(1);
}
console.log("Quick Sort:", quickSort(nums));
console.log("Merge Sort:", mergeSort(nums));

File: src/merge-sort/merge-sort.rs
fn quick_sort(mut arr: Vec<i32>) -> Vec<i32> {
    if arr.len() <= 1 { return arr; }
    let pivot = arr[arr.len()/2];
    let left: Vec<i32> = arr.iter().cloned().filter(|&x| x < pivot).collect();
    let mid: Vec<i32> = arr.iter().cloned().filter(|&x| x == pivot).collect();
    let right: Vec<i32> = arr.iter().cloned().filter(|&x| x > pivot).collect();
    [quick_sort(left), mid, quick_sort(right)].concat()
}

fn merge_sort(arr: Vec<i32>) -> Vec<i32> {
    if arr.len() <= 1 { return arr; }
    let mid = arr.len()/2;
    let left = merge_sort(arr[..mid].to_vec());
    let right = merge_sort(arr[mid..].to_vec());
    merge(left, right)
}

fn merge(mut left: Vec<i32>, mut right: Vec<i32>) -> Vec<i32> {
    let mut res = Vec::new();
    while !left.is_empty() && !right.is_empty() {
        if left[0] <= right[0] { res.push(left.remove(0)); }
        else { res.push(right.remove(0)); }
    }
    res.extend(left); res.extend(right);
    res
}

fn main() {
    let nums: Vec<i32> = std::env::args().skip(1)
        .filter_map(|s| s.parse::<i32>().ok()).collect();
    println!("Quick Sort: {:?}", quick_sort(nums.clone()));
    println!("Merge Sort: {:?}", merge_sort(nums));
}
